falcon/testing/client.py

```diff
@@ -736,6 +751,15 @@ async def _simulate_request_asgi(
             overrides `body` and sets the Content-Type header to
             ``'application/json'``, overriding any value specified by either
             the `content_type` or `headers` arguments.
+        files(dict): same as the  files parameter in requests,
```

Comment:
I find interesting but probably somewhat unintuitive that the `json` parameter is used to pass non-file parts, this isn't really obvious from the proposed documentation either. I know it is partially my fault when I think I suggested to follow `requests` somewhere in the issue, there apparently you need to pass non-file parts as `data`. I think `httpx` also follows this interface.

So I think here we have two viable options. Either try to follow the same interface, and implement `data` too (that would result in an URL-encoded form in the absence of `files`), or break with the tradition, and try to merge both into one parameter, e.g., `form`. (Or just implement `files` to get something started, and create a follow-up issue for non-file fields and URL-encoded forms? :thinking:)

Thoughts @TigreModerata @CaselIT @kgriffs?

Reply:
@CaselIT do you have any input here?

I'm leaning towards remodeling our interface to follow the popular `requests` (and `httpx`) since it is what probably the absolute majority of developers are well familiar with.
Which implies:
* Add support for `data` as in `requests` -- if provided as a `dict` or an iterable of items, then if `files` are provided, we use it as in `requests` (or in lieu of `json` in your current code).
* As in `requests`, if `files` are absent, then use `data` items to URL-encode the `POST` payload.
* As in `requests`, if `data` is a string or a bytestring, then in the same way as `body` currently is.
* Providing string `data` and `files` together either results in a `ValueError`, or we clearly describe the precedence (i.e. what overrides what, as with `json` vs `body`).
* We keep `body` as a deprecated alias for `data`.

What do you think yourself @TigreModerata? I can help, but all this would still increase the amount of work for you here :sweat_smile:

Reply:
Absolutely agree regarding the precedence, and in general, I meant to say we could follow the `requests`' suite where wise and applicable, not blindly. :slightly_smiling_face:

Reply:
Taking a look at this again, maybe we should after all diverge from `requests` here? Splitting `files` and `data` looks really clunky...

How about just adding `form` instead?
* If `form` is only a simple key-value dict, serialize using urlencoded
* If `form` has file tuples, use a multipart form
* It would be possible to specify content-type in conjunction w/ `form` where applicable, i.e., it would be possible to supply a custom `boundary` this way, or force multipart on a dict

Reply:
not having the ability of passing bytes in `data` will break people code though.

if you meant using `form` in addition to `data` we would need to have a specific rule for how they are handled.
same for `json`

Reply:
I was thinking we wouldn't touch anything existing at all, just add a `form` dict for both types of fields. And yes, document the precedence order, same as for `json`. Since `json` existed before, we can maybe make `form` even lower prio.

Reply:
so no explicit error if two of them are provided and one would be ignored?

Reply:
We can raise an error, I'm not opinionated about this. OTOH, currently `json` overrides silently, so maybe it makes sense to continue the same pattern?